Source/WebCore/animation/CSSPropertyAnimation.cpp

```diff
@@ -1695,57 +1695,89 @@ class PropertyWrapperShadow final : public AnimationPropertyWrapperBase {
+
+        return a.colorResolvingCurrentColor(fromStyleColor) == b.colorResolvingCurrentColor(toStyleColor);
     }
 
     void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const override
     {
-        auto fromStyleColor = unresolvedValue(from);
-        auto toStyleColor = unresolvedValue(to);
+        auto& fromStyleColor = value(from);
+        auto& toStyleColor = value(to);
 
         // We don't animate on currentcolor-only transition.
         // https://github.com/WebKit/WebKit/blob/main/LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/currentcolor-animation-001.html#L27
-        if (RenderStyle::isCurrentColor(fromStyleColor) && RenderStyle::isCurrentColor(toStyleColor))
+        if (fromStyleColor.isCurrentColor() && toStyleColor.isCurrentColor())
             return;
 
-        auto fromColor = value(from);
-        auto toColor = value(to);
+        auto fromColor = from.colorResolvingCurrentColor(fromStyleColor);
+        auto toColor = to.colorResolvingCurrentColor(toStyleColor);
+
         auto result = blendFunc(fromColor, toColor, context);
         (destination.*m_setter)(result);
     }
 
+#if !LOG_DISABLED
+    void logBlend(const RenderStyle& from, const RenderStyle& to, const RenderStyle& destination, double progress) const final
+    {
+        // FIXME: better logging.
+        LOG_WITH_STREAM(Animations, stream << "  blending " << property() << " from " << value(from) << " to " << value(to) << " at " << TextStream::FormatNumberRespectingIntegers(progress) << " -> " << value(destination));
+    }
+#endif
 private:
-    Color value(const RenderStyle& style) const
+    const StyleColor& value(const RenderStyle& style) const
     {
-        auto color = (style.*m_getter)();
-        return style.colorResolvingCurrentColor(color);
+        return (style.*m_getter)();
     }
 
-    StyleColor unresolvedValue(const RenderStyle& style) const
+    const StyleColor& (RenderStyle::*m_getter)() const;
+    void (RenderStyle::*m_setter)(const StyleColor&);
+};
+
+class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    PropertyWrapperColor(CSSPropertyID property, const Color& (RenderStyle::*getter)() const, void (RenderStyle::*setter)(const Color&))
+        : AnimationPropertyWrapperBase(property)
+        , m_getter(getter)
+        , m_setter(setter)
     {
-        return (style.*m_getter)();
+    }
+
+    bool equals(const RenderStyle& a, const RenderStyle& b) const override
+    {
+        if (&a == &b)
+            return true;
+
+        return value(a) == value(b);
+    }
+
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const override
+    {
+        auto result = blendFunc(value(from), value(to), context);
+        (destination.*m_setter)(result);
```

Comment:
Bit sad that we can't move this. I guess it gets copied down the line?

Reply:
I think we can.